File: advent_python/day03.py
INPUT = ["..##.......",
         "#...#...#..",
         ".#....#..#.",
         "..#.#...#.#",
         ".#...##..#.",
         "..#.##.....",
         ".#.#.#....#",
         ".#........#",
         "#.##...#...",
         "#...##....#",
         ".#..#...#.#"]

SLOPES = [(1,1), (3,1), (5,1), (7,1), (1,2)]

def traverse_forrest(inp, steps_right=3, steps_down=1):
    # (row, col)
    start = (0,0)
    trees = 0

    while start[0] < (len(inp) - 1):
        row = start[0] + steps_down
        col = start[1] + steps_right
        if col > (len(inp[row]) - 1):
            col = col - len(inp[row])
        start = (row, col)
        if inp[row][col] == '#':
            trees += 1

    return trees

assert traverse_forrest(inp=INPUT) == 7

def multiply_trees(input, slopes):
    trees = []
    for slope in slopes:
        trees.append(traverse_forrest(inp=input, steps_right=slope[0], steps_down=slope[1]))
    mul = 1
    for tree in trees:
        mul *= tree
    return mul

assert multiply_trees(input=INPUT, slopes=SLOPES) == 336

with open('../inputs/day03.txt', 'r') as f:
    data = [l.strip() for l in f]
    print(multiply_trees(input=data, slopes=SLOPES))

File: advent_cpp/day03.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <sstream>
#include <cassert>


std::string INPUT = R"(..##.......
#...#...#..
.#....#..#.
..#.#...#.#
.#...##..#.
..#.##.....
.#.#.#....#
.#........#
#.##...#...
#...##....#
.#..#...#.#)";

std::vector<std::vector<int>> STEP_SIZES {
    {1,1},
    {3,1},
    {5,1},
    {7,1},
    {1,2}
};


std::vector<std::string> mlstr2vec(std::string in) 
{
    std::vector<std::string> input;
    while (in.find("\n") < std::string::npos) {
        input.push_back(in.substr(0, in.find("\n")));
        in = in.substr((in.find("\n") + 1));
    }
    input.push_back(in);
    return input;
}

std::vector<std::string> read_input(std::string & fileName)
{
    std::vector<std::string> forrest;
    std::ifstream in(fileName.c_str());

    // Check if the object is valid
    if (!in) {
        std::cerr << "Cannot open the file: " << fileName << std::endl;
    }

    std::string str;
    // Read in the file, line by line
    while (std::getline(in, str)) 
    {
        if (str.size() > 0)
            forrest.push_back(str);
    }
    // Close the file
    in.close();

    return forrest;
}

int traverse_forrest(std::vector<std::string> forrest, int right=3, int down=1) 
{
    int x = 0;
    int trees = 0;
    int max_x = forrest[0].size();
    for (size_t y = 0; y < forrest.size(); y += down) 
    {
        char pos = forrest[y][x];
        if (pos == '#') 
            trees++;

        x += right;
        if (x > (max_x - 1)) {
            x = x - max_x;
        }
        

    }
    return trees;
}

long multiply_trees(std::vector<std::string> forrest, std::vector<std::vector<int>> step_sizes)
{
    std::vector<int> trees;
    for (std::vector<int> step : step_sizes) {
        trees.push_back(traverse_forrest(forrest, step[0], step[1]));
    }

    long mul = 1;
    for (int t: trees) {
        mul *= t;
    }
    return mul;
}


int main() {
    std::string fileName {"../inputs/day03.txt"};
    std::vector<std::string> forrest = read_input(fileName);

    assert(traverse_forrest(mlstr2vec(INPUT)) == 7);
    
    std::cout << traverse_forrest(forrest) << std::endl;

    long mul = multiply_trees(forrest, STEP_SIZES);
    std::cout << mul << std::endl;

    return 0;
}